proposals/binary-compat-only.md

```diff
@@ -0,0 +1,123 @@
+> 
+> - **If `T` is marked with `BinaryCompatOnlyAttribute`, the accessibility domain of `T` is completely inaccessible to the program text of `P` and any program that references `P`.**
+> - If the declared accessibility of `T` is public, the accessibility domain of `T` is the program text of `P` and any program that references `P`.
+> - If the declared accessibility of `T` is internal, the accessibility domain of `T` is the program text of `P`.
+> 
+> *Note*: From these definitions, it follows that the accessibility domain of a top-level unbound type is always at least the program text of the program in which that type is declared. *end note*
+> 
+> The accessibility domain for a constructed type `T<A₁, ..., Aₑ>` is the intersection of the accessibility domain of the unbound generic type `T` and the accessibility domains of the type arguments `A₁, ..., Aₑ`.
+> 
+> The accessibility domain of a nested member `M` declared in a type `T` within a program `P`, is defined as follows (noting that `M` itself might possibly be a type):
+> 
+> - **If `M` is marked with `BinaryCompatOnlyAttribute`, the accessibility domain of `M` is completely inaccessible to the program text of `P` and any program that references `P`.**
+> - If the declared accessibility of `M` is `public`, the accessibility domain of `M` is the accessibility domain of `T`.
+> - If the declared accessibility of `M` is `protected internal`, let `D` be the union of the program text of `P` and the program text of any type derived from `T`, which is declared outside `P`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `private protected`, let `D` be the intersection of the program text of `P` and the program text of `T` and any type derived from `T`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `protected`, let `D` be the union of the program text of `T`and the program text of any type derived from `T`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `internal`, the accessibility domain of `M` is the intersection of the accessibility domain of `T` with the program text of `P`.
+> - If the declared accessibility of `M` is `private`, the accessibility domain of `M` is the program text of `T`.
+
+The goal of these additions is to make it so that members marked with `BinaryCompatOnlyAttribute` are completely inaccessible to any location, they will
+not participate in member lookup, and cannot affect the rest of the program. Consequentely, this means they cannot implement interface members, they cannot
+call each other, and they cannot be overridden (virtual methods), hidden, or implemented (interface members). Whether this is too strict is the subject of
+several open questions below.
```

Comment:
The interface member part will lead to some interesting situations. Consider the following as a concret example: 

```csharp
// assembly1:
public abstract class C {
  public void Dispose() { }
}

// assembly2:
public class D : C, IDisposable { } 
```

Now lets say in version 2 the author of `assembly1` marks `C.Dispose` with `[BinaryCompatOnly]`. Execution of `assembly2` will continue to work fine because `C.Dispose` still satisfies `IDisposable.Dispose` by CLR rules. The moment the author moves to a new version of `assembly1` they will get a compilation error and be forced to re-implement `IDisposable.Dispose`. That overall may be okay but it seems like a potential pitfall in some cases.